### ISSUE TITLE
Only cookies being caught

### ISSUE THREAD

**Issue by Genus-learner** [darkarp/chromepass#39]
Ok, I forgot in my last post. Part of why I thought the memory error was real was because the server console gives this error and I receive no data
![image](https://user-images.githubusercontent.com/83723816/129410747-dec14c2e-799f-43a0-9e09-89491519f56f.png)

**Comment by darkarp** [comment]
Please download Chromepass again and try it out. I actually noticed that bug before and I think it's fixed now.

**Comment by darkarp** [comment]
This particular bug appeared because the program tried to create a directory inside `data`  but `data` didn't exist. It should be fixed :D

**Comment by Genus-learner** [comment]
Redownloaded it and even manually created a data
![image](https://user-images.githubusercontent.com/83723816/129436672-2479ccdb-1ae3-41b3-81ba-f8de851aeba9.png)
still get the same error

**Comment by darkarp** [comment]
I reproduced the error. Please check that the server has permission to write to the directory it's in. The server writes all of its data in the same directory and if it can't do so it will fail on the next writes. I would test putting the server on the Desktop since it should always have write permission there.

**Comment by Genus-learner** [comment]
Ran as admin and that indeed did fix it.
Unfortunately I am having [the same problem's as I did with the old one](https://github.com/darkarp/chromepass/issues/36):
![image](https://user-images.githubusercontent.com/83723816/129460373-e7e0c938-c963-4892-bb28-fb6a09e3ae6a.png)

It is only getting cooking.
Not only that but the server isn't actually writing the data to any files
I had the same issue with your old version but I was able to get it to write data by running the python server. But there's no option to make a python version of the server in the latest iteration.

In short:
-Still only getting cookies
-Exe version of the Server doesn't create data files like the python server did in the old version

**Comment by darkarp** [comment]
Huh, that's weird! It's possible there's a variable I'm not accounting for as well.
A couple suggestions:
1. Do you have chrome open during this? (or anything else?). Maybe try closing everything,  just in case (is something preventing access to that particular file?).
2. Maybe turn off AVs/Defender, just to eliminate that possibility (maybe related to point 1).
3. Run the client as administrator, in case perhaps there are some permission problems with accessing that file specifically?

**Comment by Genus-learner** [comment]
Thank you for your quick responses!
I'll try these out soon and let you know how it goes

**Comment by darkarp** [comment]
I remembered another potential problem that I had thought about before and ultimately forgot.
Perhaps the login data file is located elsewhere and not in the Default folder. 
For those machines where it doesn't work, could you check the files in `%LOCALAPPDATAA%/Google/Chrome/User Data/Default`?
There should be a Cookies and Login Data folder inside.

**Comment by Genus-learner** [comment]
@ C:\Users\Username\AppData\Local\Google\Chrome\User Data\Default
There is login data

![image](https://user-images.githubusercontent.com/83723816/129784522-86c3fe4c-ac26-4323-b6a6-f32f905d82c3.png)

**Comment by Genus-learner** [comment]
> Huh, that's weird! It's possible there's a variable I'm not accounting for as well.
> A couple suggestions:
> 
> 1. Do you have chrome open during this when running the client? (or anything else?). Maybe try closing everything,  just in case (is something preventing access to that particular file?).
> 2. Maybe turn off AVs/Defender, just to eliminate that possibility (maybe related to point 1).
> 3. Run the client as administrator, in case perhaps there are some permission problems with accessing that file specifically?
> 
> If any of these work maybe we can narrow it down further.
> Thanks for your patience!

I tried your suggestions. No improvement.

**Comment by darkarp** [comment]
This is really strange. Would it be okay if I gave you a version of chromepass in form of an executable to test it out? I'm assuming you're testing it on a VM

**Comment by darkarp** [comment]
[chromepass.zip](https://github.com/darkarp/chromepass/files/7004023/chromepass.zip)
It's just an executable that's pointint to 127.0.0.1:80. Could you test it out by running it with powershell/cmd? That way you can see what it prints out :D

**Comment by Genus-learner** [comment]
> This is really strange. Would it be okay if I gave you a version of chromepass in form of an executable to test it out? I'm assuming you're testing it on a VM

Chromepass is installed on a VM, but the client I am testing on my actual computer.


> [chromepass.zip](https://github.com/darkarp/chromepass/files/7004023/chromepass.zip)
> It's just an executable that's pointing to 127.0.0.1:80. Could you test it out by running it with powershell/cmd? That way you can see what it prints out and helps us figure out where the problem is more exactly :D
Tested the file you sent on the VM
![image](https://user-images.githubusercontent.com/83723816/130158866-8bdcce64-feac-41ee-89a8-d5395411cba9.png)

**Comment by darkarp** [comment]
Was the server running at the time the client was executed?

**Comment by darkarp** [comment]
by the way, it's totally safe to test that even on your computer, it just does the same thing as a client generated with the default 127.0.0.1:80 configuration sending to the local server, but with a lot of different printing along the way :D

**Comment by Genus-learner** [comment]
This is the printing on the vm where it is able to get login and cookies
![image](https://user-images.githubusercontent.com/83723816/130238320-81e4bd1c-4aa1-4735-8485-b98c0cd67c6d.png)

This is the printing on my actual computer where it cant get login info but can get cookies
![image](https://user-images.githubusercontent.com/83723816/130239125-17ed51c8-1767-439a-9138-887792d9c9a8.png)

Server was running this time in both scenarios

**Comment by darkarp** [comment]
That's perfect, now we know it's failing at the decryption of the password in login. I think I might know why this is the case. 
I have just pushed an update to chromepass, could you try generating the client now to figure out if it's working?

**Comment by Genus-learner** [comment]
> That's perfect, now we know it's failing at the decryption of the password in login. I think I might know why this is the case.
> I have just pushed an update to chromepass, could you try generating the client now to figure out if it's working?

Latest build can't make a client for some reason
Didnt happen with the previous ones
It feels like I'm just doing something wrong at this point lol

**Comment by darkarp** [comment]
hmm sorry about that, it might be a problem on my end. Do you see any error message inside the `logs.log` folder? (maybe it would be best to delete that file and try to build the client again, to remove the old errors)

**Comment by Genus-learner** [comment]
Here's the end of the log
![image](https://user-images.githubusercontent.com/83723816/130337211-acbf60d9-b839-466e-8437-114b723061d7.png)

Here's the full log file in case I missed something important

> DEBUG:root:Compiling proc-macro2 v1.0.28
> DEBUG:root:Compiling unicode-xid v0.2.2
> DEBUG:root:Compiling syn v1.0.74
> DEBUG:root:Compiling winapi v0.3.9
> DEBUG:root:Compiling autocfg v1.0.1
> DEBUG:root:Compiling cfg-if v1.0.0
> DEBUG:root:Compiling winapi-build v0.1.1
> DEBUG:root:Compiling memchr v2.4.0
> DEBUG:root:Compiling futures-core v0.3.16
> DEBUG:root:Compiling log v0.4.14
> DEBUG:root:Compiling version_check v0.9.3
> DEBUG:root:Compiling proc-macro-hack v0.5.19
> DEBUG:root:Compiling winapi v0.2.8
> DEBUG:root:Compiling futures-sink v0.3.16
> DEBUG:root:Compiling lazy_static v1.4.0
> DEBUG:root:Compiling slab v0.4.4
> DEBUG:root:Compiling futures-channel v0.3.16
> DEBUG:root:Compiling cfg-if v0.1.10
> DEBUG:root:Compiling futures-task v0.3.16
> DEBUG:root:Compiling pin-project-lite v0.2.7
> DEBUG:root:Compiling proc-macro-nested v0.1.7
> DEBUG:root:Compiling iovec v0.1.4
> DEBUG:root:Compiling libc v0.2.99
> DEBUG:root:Compiling futures-io v0.3.16
> DEBUG:root:Compiling pin-utils v0.1.0
> DEBUG:root:Compiling pin-project-lite v0.1.12
> DEBUG:root:Compiling bytes v0.5.6
> DEBUG:root:Compiling serde_derive v1.0.127
> DEBUG:root:Compiling serde v1.0.127
> DEBUG:root:Compiling smallvec v1.6.1
> DEBUG:root:Compiling pin-project-internal v0.4.28
> DEBUG:root:Compiling scopeguard v1.1.0
> DEBUG:root:Compiling getrandom v0.1.16
> DEBUG:root:Compiling matches v0.1.8
> DEBUG:root:Compiling convert_case v0.4.0
> DEBUG:root:Compiling percent-encoding v2.1.0
> DEBUG:root:Compiling tinyvec_macros v0.1.0
> DEBUG:root:Compiling copyless v0.1.5
> DEBUG:root:Compiling itoa v0.4.7
> DEBUG:root:Compiling bitflags v1.3.1
> DEBUG:root:Compiling typenum v1.13.0
> DEBUG:root:Compiling unicode-segmentation v1.8.0
> DEBUG:root:Compiling either v1.6.1
> DEBUG:root:Compiling async-trait v0.1.51
> DEBUG:root:Compiling ppv-lite86 v0.2.10
> DEBUG:root:Compiling bytes v1.0.1
> DEBUG:root:Compiling match_cfg v0.1.0
> DEBUG:root:Compiling fnv v1.0.7
> DEBUG:root:Compiling ryu v1.0.5
> DEBUG:root:Compiling ipconfig v0.2.2
> DEBUG:root:Compiling const_fn v0.4.8
> DEBUG:root:Compiling linked-hash-map v0.5.4
> DEBUG:root:Compiling widestring v0.4.3
> DEBUG:root:Compiling quick-error v1.2.3
> DEBUG:root:Compiling serde_json v1.0.66
> DEBUG:root:Compiling regex-syntax v0.6.25
> DEBUG:root:Compiling httparse v1.4.1
> DEBUG:root:Compiling encoding_rs v0.8.28
> DEBUG:root:Compiling hashbrown v0.11.2
> DEBUG:root:Compiling byteorder v1.4.3
> DEBUG:root:Compiling cpufeatures v0.1.5
> DEBUG:root:Compiling opaque-debug v0.3.0
> DEBUG:root:Compiling language-tags v0.2.2
> DEBUG:root:Compiling mio-uds v0.6.8
> DEBUG:root:Compiling mime v0.3.16
> DEBUG:root:Compiling base64 v0.13.0
> DEBUG:root:Compiling instant v0.1.10
> DEBUG:root:Compiling ws2_32-sys v0.2.1
> DEBUG:root:Compiling kernel32-sys v0.2.2
> DEBUG:root:Compiling futures-macro v0.3.16
> DEBUG:root:Compiling futures-util v0.3.16
> DEBUG:root:Compiling indexmap v1.7.0
> DEBUG:root:Compiling num-traits v0.2.14
> DEBUG:root:Compiling num-integer v0.1.44
> DEBUG:root:Compiling standback v0.2.17
> DEBUG:root:Compiling generic-array v0.14.4
> DEBUG:root:Compiling time v0.2.27
> DEBUG:root:Compiling cookie v0.14.4
> DEBUG:root:Compiling tracing-core v0.1.18
> DEBUG:root:Compiling lock_api v0.4.4
> DEBUG:root:Compiling unicode-bidi v0.3.5
> DEBUG:root:Compiling form_urlencoded v1.0.1
> DEBUG:root:Compiling tinyvec v1.3.1
> DEBUG:root:Compiling heck v0.3.3
> DEBUG:root:Compiling http v0.2.4
> DEBUG:root:Compiling bytestring v1.0.0
> DEBUG:root:Compiling lru-cache v0.1.2
> DEBUG:root:Compiling fxhash v0.2.1
> DEBUG:root:Compiling unicode-normalization v0.1.19
> DEBUG:root:Compiling quote v1.0.9
> DEBUG:root:Compiling net2 v0.2.37
> DEBUG:root:Compiling parking_lot_core v0.8.3
> DEBUG:root:Compiling socket2 v0.3.19
> DEBUG:root:Compiling hostname v0.3.1
> DEBUG:root:Compiling winreg v0.6.2
> DEBUG:root:Compiling aho-corasick v0.7.18
> DEBUG:root:Compiling tracing v0.1.26
> DEBUG:root:Compiling num_cpus v1.13.0
> DEBUG:root:Compiling time v0.1.44
> DEBUG:root:Compiling rand_core v0.5.1
> DEBUG:root:Compiling idna v0.2.3
> DEBUG:root:Compiling miow v0.2.2
> DEBUG:root:Compiling parking_lot v0.11.1
> DEBUG:root:Compiling resolv-conf v0.7.0
> DEBUG:root:Compiling regex v1.5.4
> DEBUG:root:Compiling threadpool v1.8.1
> DEBUG:root:Compiling rand_chacha v0.2.2
> DEBUG:root:Compiling block-buffer v0.9.0
> DEBUG:root:Compiling digest v0.9.0
> DEBUG:root:Compiling url v2.2.2
> DEBUG:root:Compiling mio v0.6.23
> DEBUG:root:Compiling chrono v0.4.19
> DEBUG:root:Compiling rand v0.7.3
> DEBUG:root:Compiling sha-1 v0.9.7
> DEBUG:root:Compiling derive_more v0.99.16
> DEBUG:root:Compiling actix-macros v0.1.3
> DEBUG:root:Compiling thiserror-impl v1.0.26
> DEBUG:root:Compiling enum-as-inner v0.3.3
> DEBUG:root:Compiling time-macros-impl v0.1.2
> DEBUG:root:Compiling pin-project-internal v1.0.8
> DEBUG:root:Compiling actix-web-codegen v0.4.0
> DEBUG:root:Compiling tokio v0.2.25
> DEBUG:root:Compiling pin-project v0.4.28
> DEBUG:root:Compiling thiserror v1.0.26
> DEBUG:root:Compiling actix-threadpool v0.3.3
> DEBUG:root:Compiling time-macros v0.1.1
> DEBUG:root:Compiling pin-project v1.0.8
> DEBUG:root:Compiling tokio-util v0.3.1
> DEBUG:root:Compiling actix-service v1.0.6
> DEBUG:root:Compiling futures v0.3.16
> DEBUG:root:Compiling serde_urlencoded v0.7.0
> DEBUG:root:Compiling actix-router v0.2.7
> DEBUG:root:Compiling actix-rt v1.1.1
> DEBUG:root:Compiling tracing-futures v0.2.5
> DEBUG:root:Compiling actix-codec v0.3.0
> DEBUG:root:Compiling toml v0.5.8
> DEBUG:root:Compiling trust-dns-proto v0.19.7
> DEBUG:root:Compiling h2 v0.2.7
> DEBUG:root:Compiling actix-utils v2.0.0
> DEBUG:root:Compiling winres v0.1.11
> DEBUG:root:Compiling trust-dns-resolver v0.19.7
> DEBUG:root:Compiling actix-server v1.0.4
> DEBUG:root:Compiling actix-tls v2.0.0
> DEBUG:root:Compiling chromepass-server v0.1.0 (C:\Windows\system32\chromepass\templates\chromepass-server)
> DEBUG:root:Compiling actix-connect v2.0.0
> DEBUG:root:Compiling actix-testing v1.0.1
> DEBUG:root:Compiling actix-http v2.2.1
> DEBUG:root:Compiling awc v2.0.3
> DEBUG:root:Compiling actix-web v3.3.2
> DEBUG:root:Finished release [optimized] target(s) in 9m 50s
> DEBUG:root:error: failed to parse manifest at `C:\Windows\system32\chromepass\templates\chromepass-build\Cargo.toml`
> DEBUG:root:
> DEBUG:root:Caused by:
> DEBUG:root:the cargo feature `strip` requires a nightly version of Cargo, but this is the `stable` channel
> DEBUG:root:See https://doc.rust-lang.org/book/appendix-07-nightly-rust.html for more information about Rust release channels.
> DEBUG:root:See https://doc.rust-lang.org/cargo/reference/unstable.html#profile-strip-option for more information about using this feature.
> DEBUG:root:Compiling chromepass-server v0.1.0 (C:\Windows\system32\chromepass\templates\chromepass-server)
> DEBUG:root:Finished release [optimized] target(s) in 41.40s
> DEBUG:root:error: failed to parse manifest at `C:\Windows\system32\chromepass\templates\chromepass-build\Cargo.toml`
> DEBUG:root:
> DEBUG:root:Caused by:
> DEBUG:root:the cargo feature `strip` requires a nightly version of Cargo, but this is the `stable` channel
> DEBUG:root:See https://doc.rust-lang.org/book/appendix-07-nightly-rust.html for more information about Rust release channels.
> DEBUG:root:See https://doc.rust-lang.org/cargo/reference/unstable.html#profile-strip-option for more information about using this feature.
> DEBUG:root:Compiling chromepass-server v0.1.0 (C:\users\user\desktop\chromepass\templates\chromepass-server)
> DEBUG:root:Finished release [optimized] target(s) in 25.68s
> DEBUG:root:error: failed to parse manifest at `C:\users\user\desktop\chromepass\templates\chromepass-build\Cargo.toml`
> DEBUG:root:
> DEBUG:root:Caused by:
> DEBUG:root:the cargo feature `strip` requires a nightly version of Cargo, but this is the `stable` channel
> DEBUG:root:See https://doc.rust-lang.org/book/appendix-07-nightly-rust.html for more information about Rust release channels.
> DEBUG:root:See https://doc.rust-lang.org/cargo/reference/unstable.html#profile-strip-option for more information about using this feature.

**Comment by Genus-learner** [comment]
>Meanwhile, I pushed another update, which just makes the error messages a bit more descriptive so we can know what's wrong :)

I'll take a look

**Comment by darkarp** [comment]
My apologies. I just checked the error message. It was a line I added to the config file. I just fixed it. It should build.

**Comment by Genus-learner** [comment]
> My apologies. I just checked the error message. It was a line I added to the config file. I just fixed it. It should build.

![image](https://user-images.githubusercontent.com/83723816/130337800-7d89f3c8-0c6b-4271-99ed-f8dfe2f3dad2.png)

**Comment by darkarp** [comment]
Looking into it! Again, thanks for the patience!!

**Comment by darkarp** [comment]
Fortunately it seems that issue is related to a specific dependency which isn't used anymore in my project so I just removed it. Hopefully it works now :D (just updated it)

**Comment by Genus-learner** [comment]
I got a clean build, but now the server isn't receiving anything from the client?
I ran an old version of the client and it seemed to work (but since its an old version of course it still doesn't get any login data)

**Comment by darkarp** [comment]
Pushed another update. I noticed a few issues that might be causing your current problem, mainly the way that the program figures out where the local appdata program is, I realized not all systems will have the required environment variable so it now uses a different method. I was also checking whether the decryption key was correctly found or not, which also makes the script not work for earlier versions of Chromium-based browsers. (It would also not be able to decrypt passwords saved from when the browser was in its older version, even if updated afterwards).

It should work now, I hope :D

> lol hopefully all this troubleshooting will make everything go super smoothly for anyone else who wants to learn how this all works in the future
Yeah hopefully :P

**Comment by Genus-learner** [comment]
I would like to congratulate you sir, as of right now, it is working on both the VM and the actual computer that it wasn't working on when this thread was created (and in #36 ). When I get the chance I'll test it on my other computers to make sure it still works in them after all these changes that have happened.

> I noticed a few issues that might be causing your current problem, mainly the way that the program figures out where the local appdata program is, I realized not all systems will have the required environment variable so it now uses a different method
What has been changed in the decryption process since the beginning of the thread? What was the main issue causing it to not get login data?

**Comment by darkarp** [comment]
Thank you! I'd like to thank you as well for helping track this issue down, it wouldn't have been possible without you! 
I'll wait until I close the issue to make sure it works on your other computers as well, but I'll mark it as complete meanwhile :D

**Comment by Genus-learner** [comment]
> I'll wait until I close the issue to make sure it works on your other computers as well, but I'll mark it as complete meanwhile :D

Good news!
it seems to be working with all my computers.
I think you officially fixed it!


> I noticed a few issues that might be causing your current problem, mainly the way that the program figures out where the local appdata program is, I realized not all systems will have the required environment variable so it now uses a different method

What has been changed in the decryption process since the beginning of the thread? What was the main issue causing it to not get login data?

**Comment by darkarp** [comment]
That's awesome!

So a couple of things got changed. Before, the decryption process went like this:
1. Look in the Local State file for the AES256-gcm key used to decrypt passwords in Chrome 80+
2. The key itself is encrypted with DPAPI so we decrypt the key.
3. Then we look for the Login Data and Cookie directory and use the decrypted key to get the plaintext passwords.

Initially I thought the problem was only on step 3, and it partially was. Chrome version 80 and above store the passwords in this encryption format (encrypted by the aes256-gcm key in Local State and having that key encrypted with DPAPI). Before Chrome version 80, passwords were just all encrypted with DPAPI. Now the problem is that even if you update Chrome to the latest versions, all the passwords that were saved in the Chrome<80 format will remain in that format and they won't be able to de decrypted. So I changed the decryption to simple DPAPI if the AES256-GCM fails. 

This wasn't enough however because it turns out I was also only decrypting passwords if a valid key was found in Local State, which there doesn't have to be one if the version < 80. This is more of a "making sure it works on all versions" even though most people will have updated by now. 

The last problem was the method with which I retrieved the directories. I was using LOCALAPPDATA environment variable but that may not exist. Now, I directly build the correct path from the username.

So in summary, the main issue was that the decryption failed if there were any passwords stored from Chrome versions before 80.
